moby/moby
# Import libnetwork fix for rolling updates

**- What I did**

This patch allows endpoints to complete servicing connections while being removed from a service.  The fix is mostly within libnetwork.  There is a small change to the moby code to remove a now-superfluous call to sandbox.DisableService().  However, moby would function correctly (just sub-optimally) without said removal.  (see the changelog for for the commits further details)

This fix addresses issue raised in https://github.com/docker/libnetwork/pull/2074 and is related to https://github.com/moby/moby/issues/30321.  The libnetwork PR for this fix is here https://github.com/docker/libnetwork/pull/2112.  This PR does not include any other libnetwork PRs.  The removal of the call to DisableService() from the releaseNetworks() function reverts the change introduced in https://github.com/moby/moby/pull/35960.

**- How I did it**

The fix works by initially down-weighting a container endpoint in the load balancer to 0 while keeping the endpoint present in the load balancer.  This allows traffic to continue to flow to the endpoint while preventing new connections from going to the endpoint.  This allows the container to complete requests during the "stop_grace_period" and then exit when finished without interruption of service.

This change requires propagating the status of disabled service endpoints via the networkDB.  Accordingly, the patch includes both code to generate and handle service update messages.  It also augments the service structure with a ServiceDisabled boolean to convey whether an endpoint should ultimately be removed or just disabled.  This, naturally, required a rebuild of the protocol buffer code.

The protocol buffer encoding is designed to support additions of fields to messages in a backwards-compatible manner.  Protocol buffer unmarshalling code automatically skips past any fields that it isn't
aware of.  As a result, an older moby daemon without this fix can receive and will process correctly networkDB messages from newer moby daemons with this patch.

As it turns out, the additional field is simply a bool that is otherwise irrelevent on networkDB create and delete events.  So its absence in older moby daemon processing has no impact.  However, the fix leverages the "update" networkDB message which was previously unused in libnetwork.  Although older libnetwork implementations parse the message cleanly, they will see the message as unexpected and as such issue a log at error level indicating the receipt of such.

Other than this there should be no other negative impact for use of this patch in mixed environments. (Although older mobys won't be able to gracefully downgrade connections on their nodes of course.)

Signed-off-by: Chris Telfer <ctelfer@docker.com>

**- How to verify it**

One can verify this in a swarm environment by the following steps:
  * `git clone https://github.com/ctelfer/docker-rolling-upgrade-test.git`
  * `cd docker-rolling-upgrade-test`
  * OPTIONAL: edit Makefile and docker-compose.yml to change your tag for the image you want to build
  * `make build`    -- do this on the swarm manager or with your env variables pointing to it
  * `docker stack deploy -c docker-compose.yml t`  -- likewise on the manager or w/ docker env pointing to it
  * open an ssh connection to one of the workers in the swarm
  * `curl https://LOCALIPADDR/?stream=1 > /dev/null` -- open a streaming http connection.  Make sure LOCALIPADDR is one of the IP addresses of the worker, but not localhost
  * verify from `curl` stderr output that the stream is transferring data
  * on the manager: `docker service update t_webservice --force` -- force an update
  * wait for the update to converge
  * on the worker: verify that `curl` is still passing traffic.  In the current version of moby, this will hang.  In the updated version, traffic will continue until the grace period (1 hour in my config) ends.
  * open another ssh connection to the same worker doing the curl
  * `nsenter --net=/var/run/docker/netns/ingress_sbox ipvsadm -L`
  * verify that there are 4 entries:  3 with weight 1, and 1 with weight 0
  * kill the `curl` stream with CTRL-C
  * use `nsenter ... ipvsadm -L` on the worker again and verify that it goes back to 3 entries each with weight 1
  
**- Description for the changelog**

Import libnetwork fix for rolling updates

**- A picture of a cute animal (not mandatory but encouraged)**
